### ISSUE TITLE
CALL_EXCEPTION Error: data out-of-bounds

### ISSUE THREAD

**Issue by krebernisak** [ethereum-optimism/optimism-integration#46]
Hi guys!

I am testing out the Chainlink OVM deployment and I am having issues reading from deployed contracts. 
This happens every time against my local L2 chain, and only in some cases when testing against the Goerli L2 deployment.

Repro steps:

- Checkout this repo at `7d60e2a` and run the L2 chain with the default configuration.
- Deploy a simple ERC20 token, [LinkToken](https://github.com/smartcontractkit/LinkToken/blob/feature/ovm-integration/contracts/v0.6/LinkToken.sol) in this case. [Here](https://gist.github.com/krebernisak/11fafcd9a90efda32ed6244894c4a945) is the ABI.
- Call the `await token.totalSupply()` function using `"ethers": "^5.0.7"`

This is the debug and error data received:

```
FunctionFragment {
  type: 'function',
  name: 'totalSupply',
  constant: true,
  inputs: [],
  outputs: [
    ParamType {
      name: null,
      type: 'uint256',
      indexed: null,
      components: null,
      arrayLength: null,
      arrayChildren: null,
      baseType: 'uint256',
      _isParamType: true
    }
  ],
  payable: false,
  stateMutability: 'view',
  gas: null,
  _isFragment: true
}

Bytes: Uint8Array(0) []

Error: data out-of-bounds (length=0, offset=32, code=BUFFER_OVERRUN, version=abi/5.0.7)
    at Logger.makeError (/[truncated]/node_modules/@ethersproject/logger/src.ts/index.ts:205:28)
    at Logger.throwError (/[truncated]/node_modules/@ethersproject/logger/src.ts/index.ts:217:20)
    at Reader._peekBytes (/[truncated]/node_modules/@ethersproject/abi/src.ts/coders/abstract-coder.ts:182:24)
    at Reader.readBytes (/[truncated]/node_modules/@ethersproject/abi/src.ts/coders/abstract-coder.ts:196:26)
    at Reader.readValue (/[truncated]/node_modules/@ethersproject/abi/src.ts/coders/abstract-coder.ts:203:36)
    at NumberCoder.decode (/[truncated]/node_modules/@ethersproject/abi/src.ts/coders/number.ts:44:28)
    at /[truncated]/node_modules/@ethersproject/abi/src.ts/coders/array.ts:108:31
    at Array.forEach (<anonymous>)
    at Object.unpack (/[truncated]/node_modules/@ethersproject/abi/src.ts/coders/array.ts:89:12)
    at TupleCoder.decode (/[truncated]/node_modules/@ethersproject/abi/src.ts/coders/tuple.ts:27:41) {
  reason: 'data out-of-bounds',
  code: 'BUFFER_OVERRUN',
  length: 0,
  offset: 32
}
```

Looks like the [decodeFunctionResult](https://github.com/ethers-io/ethers.js/blob/efeeda5e5c3b22f0c211783a635941dc8035ae25/packages/abi/src.ts/interface.ts#L314) fails, and `CALL_EXCEPTION` is reported back.

This happens when calling any view function against the local L2 deployment.
Please note that these contracts are tested against the [OVM Ganache provider](https://github.com/ethereum-optimism/optimism-monorepo/tree/master/packages/ovm-toolchain/src/ganache) and pass the tests with flying colors.

**EDIT: If I defer the read call for just 100 ms after the contract deployment, then the error disappears.**

**Comment by tynes** [comment]
Hey @krebernisak, are you using the `JsonRpcProvider`? Also are you waiting for the receipt before calling `totalSupply`?

**Comment by krebernisak** [comment]
Thanks, Mark!

Yes, I am using the `JsonRpcProvider`, and NO I was not waiting for the receipt so that solved the issue for this simple case.

But I still think there is some kind of an underlying issue here as I am having the same error, with some other workflows, where timing is not a factor.

I'll try to investigate some more and report back.

**Comment by tynes** [comment]
Thanks for the info Kristijan! Would really appreciate any additional information in more complex workflows. It sounds like a race condition if the receipt isn't waited on and zero bytes are being returned when it is expecting a uint256 to decode client side.

**Comment by krebernisak** [comment]
@tynes I made [a simple project](https://github.com/krebernisak/ovm-deploy-debug) where you can test this out.

If I deploy to my local network, I get the receipt with 1 as status, the block number, and everything. But there is no code at my address and of course, I can't call any functions. Timeouts don't help.

If I deploy to `goerli.optimism.io`, then it works as expected.

I have similar problems with state-changing functions. I would call a function, assuming success, but upon inspection, my state would not be updated.

**Comment by tynes** [comment]
Thanks for sharing that, will run it to help debug. Did you happen to compile the contract you are trying to deploy with the OVM compiler?
https://www.npmjs.com/package/@eth-optimism/solc

It is possible that the contract is failing the safety checker, it could be caused by an unsafe constructor.

**Comment by krebernisak** [comment]
Yeah, it's compiled using the OVM solc v0.6.12.

These are the constructor arguments:

```
536166654d6174683a206d756c7469706c69636174696f6e206f766572666c6f77000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000064000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000002a000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000095465737420342e32300000000000000000000000000000000000000000000000
```

If I try to disassemble it using the [Etherscan opcode-tool](https://etherscan.io/opcode-tool), I find some unsafe candidates:

```
[3] '4d'(Unknown Opcode)
...
[255] '2a'(Unknown Opcode)
...
[320] SLOAD
```

Could these fail the safety checker?

But if the constructor is unsafe, the question is why does it get deployed successfully to `goerli.optimism.io`?

**Comment by karlfloersch** [comment]
@krebernisak thanks for the repo to reproduce! I just tested this locally and seemingly successfully deployed! Also looked like your constructor was safe based on a quick check. Maybe `docker-compose pull` -- could have been broken on an earlier version

<img width="892" alt="Screen Shot 2020-12-16 at 9 17 27 AM" src="https://user-images.githubusercontent.com/706123/102382653-859d3100-3f7f-11eb-8be2-69fcc7a1186d.png">

**Comment by krebernisak** [comment]
You are right, looks like I did not pull the updated containers for the last day or two, and a lot can happen in that time 💯 

With the latest update, I don't have this issue anymore!
Though, it looks like I still have the issue of not being able to write to this contract after it's deployed. I'll investigate some more, and if necessary open another issue for that. Closing this one!

Thanks, @karlfloersch & @tynes for debugging this with me 🙏
